feat(posts): We love speed 2024 (#455)
BedrockStreaming/tech.bedrockstreaming.com: _posts/2024-10-29-we-love-speed-2024.md

---
layout: post
title: We love speed 2024 ❤️

> description: How did we improve the monitoring practices on the mobile service? Let's find out.

Author: impossible, le copier-coller c'est pas notre style 👉 👈

author: [ j_nginn, j_poissonnet ]

> tags: [ monitoring, production, newrelic, shared practice, alerting ]

Author: en effet 👍

color: rgb(251,87,66)
language: fr
thumbnail: "images/posts/2024-10-29-we-love-speed-2024/welovespeed_2024-1709240237.jpg"
---

Cette année, nous avons eu la chance de participer à la conférence We love speed, une conférence annuelle axée sur la
performance du web.

> Etant donné que c’est un domaine qui nous passionne <insérer le lien d’un article qui en parle ?>, nous sommes très

Author: j'ai pas trouvé d'article au final

Google récemment https://developers.google.com/search/blog/2023/05/introducing-inp?hl=fr

> L’idée de cette métrique est de se rapprocher de l’expérience de l’utilisateur envers la réactivité d’une application
> web. Cette métrique observe le temps entre une interaction utilisateur et une réponse visuelle de notre interface.

Reviewer: ```suggestion
L’objectif de cette métrique est de refléter l’expérience utilisateur en mesurant la réactivité d’une application. Elle observe le temps entre une interaction utilisateur et une réponse visuelle de notre interface.
```
Pour ne pas répéter "cette métrique", j'avais un peu de mal avec "envers" et le sens de la phrase

L’idée de cette métrique est de se rapprocher de l’expérience de l’utilisateur envers la réactivité d’une application
web. Cette métrique observe le temps entre une interaction utilisateur et une réponse visuelle de notre interface.

![Preload with surgical precision](/images/posts/2024-10-29-we-love-speed-2024/team_picture.jpeg)

> ## HTMX

Reviewer 2: Je le note pour ce talk, valable sans doute pour les suivants : moyen de mettre un lien vers la video ? ou le nom de l'auteur et un lien vers un blog ou autre ?

Et "HTMX" c'était le titre entier du talk ?

> Lors de ce talk, il nous a été présenté un outil très intéressant pour améliorer l’expérience utilisateur d’une

Reviewer 2: ```suggestion
Lors de ce talk, nous avons découvert un outil très intéressant pour améliorer l’expérience utilisateur d’une
```
La voix active ("nous avons fait"), c'est plus engageant que la voix passive ("un truc a été fait") ;-)

application web sans effectuer de grosses modifications.

> Il s'agit de HTMX, c'est une bibliothèque Javascript qui permet d’améliorer l’expérience utilisateur en ajoutant des

Reviewer 2: ```suggestion
Il s'agit de HTMX : une bibliothèque Javascript qui permet d’améliorer l’expérience utilisateur en ajoutant des
```

fonctionnalités de type SPA (Single Page Application) à une application web classique et de façon non intrusive.

> Par exemple, on peut charger un CDN de manière asynchrone, ou encore charger des images en avance.

Reviewer 2: Je ne comprends pas "charger un CDN" 🤔

C’est-a-dire que si HTMX venait à ne pas démarrer, votre application web se comporterait de la même manière, mais sans
les améliorations de temps d’interaction.

HTMX surcharge la manière dont vos liens et images vont être chargés par le navigateur.
Ainsi, lors de la prochaine interaction avec le navigateur, ce dernier sera déjà prêt à servir les ressources.
Le principe de HTMX consiste à ajouter des balises HTML spécifiques dans le DOM qu'il va lire et en déduire les
comportements à son chargement.
Cette manipulation est appelé le "DOM morphing". Grâce à ce processus le temps de chargement est réduite et on évite
l'effet "blink" (page blanche lors du chargement de la page).
Il est à noter que ces comportements ne sont qu’un embellissement proposé par HTMX, il est tout à fait possible
d’ajouter par exemple l’attribut `preload` sur une balise `a` pour demander le chargement en avance du lien par le
navigateur.

> <div style="display: flex">

> > <img src="https://ca.slack-edge.com/T108ZKPMF-U01FQRQ8FT7-dfb12b21fb0d-192" alt="Julie" style="padding: 0;border-radius: 50%; height: 70px; margin: 10px">

Reviewer 2: Ca rend pas mal ces blocs où vous donnez vos avis perso, c'est top 👍 

Sur celui-ci : je linkerais vers une image que nous hébergeons, au lieu de hotlinker vers slack.

> Comme nous utilisons React pour notre application, l'utilisation de HTMX n'est pas vraiment utile.
> Il est déjà possible avec React de précharger les ressources en avance. Mais ça reste un outil intéressant...
> </div>

> <div style="display: flex">
> ...effectivement, HTMX semble être intéressant, mais on se retrouve à ajouter
> beaucoup d'attributs dans le HTML. Ça peut le rendre le markup moins lisible. Et en plus, ça donne l'impression de recoder les comportements du navigateur.
> <img src="/images/avatar/j_poissonnet.jpg" alt="Jules" style="padding: 0;border-radius: 50%; height: 70px; margin: 10px">
> </div>

## INP vs JS

Le deuxième talk a une place particulière dans notre cœur ❤️ puisqu’il a été donné par notre cher Jean-Pierre Vincent,
qui a audité les performances du web de Bedrock, il y a deux ans.
Lors de ce talk, Jean-Pierre nous a donné la feuille de route pour éviter au mieux la déferlante de Javascript que vos
utilisateurs reçoivent au chargement de votre site.

![JS Tsunami storming your users](/images/posts/2024-10-29-we-love-speed-2024/js_tsunami.jpeg)

> L'INP (Interaction to Next Paint) est une métrique qui mesure le temps entre une interaction utilisateur et le prochain.

Reviewer 2: J'ai l'impression qu'il manque un mot après "le prochain" (sinon, on croit que c'est "le prochain utilisateur", vu que "utilisateur" est le dernier mot employé)

> L’idée générale, c'est de pouvoir mesurer l’incapacité du navigateur à réagir. Après avoir récupéré des mesures, il est

Reviewer 2: ```suggestion
L’idée générale est de pouvoir mesurer l’incapacité du navigateur à réagir. Après avoir récupéré des mesures, il est
```

bon de

> se rappeler qu’il y a un biais de selection pour les données de Crux. En effet, il n’est calculé que sur les appareils

Author: Oui c'est vrai que ça fait pas de mal de rajouter un peu de contexte

Google (c’est le principe). Une fois qu’on a récolté des métriques de performance de nos utilisateurs, si on veut
travailler sur notre site web et avoir une bonne idée du ressenti de nos utilisateurs, l’idéal est de tester avec un

> véritable Samsung S8 par exemple.

Author: En fait dans les conseils de la conf, JP nous disait qu'un S8 est un téléphone qui a beaucoup été vendu et qui à l'heure actuelle représente plutot fidelement l'utilisateur moyen. Je pense que ce n'est pas que c'est mieux mais juste que c'est un appareil plus représentatif

> L’INP est une métrique qui peut être influencée par des interactions qui ne sont pas prévues par les devs. C’est
> pourquoi il est important de se baser sur des données utilisateurs, car on est très souvent étonné de voir que quand les
> temps de chargement sont un poil trop long à leur goût, ils se mettent à cliquer partout 🤷

Reviewer: ```suggestion
L’INP est une métrique qui peut être influencée par des interactions qui ne sont pas prévues par les devs. Par exemple, on a été étonnés de constater que lorsque les temps de chargement sont un poil trop longs à leur goût, les utilisateurs se mettent à cliquer partout 🤷 C’est pourquoi il est important de se baser sur des données réelles. 
```
J'ai mis la conclusion à la fin, peut-être que moi mais j'ai eu un peu de mal à comprendre à la première lecture

Author: Le fait que t'aies pas compris du premier coup montre surement un besoin de reformulation 😉

![INP est bousculé par la charge de js!](/images/posts/2024-10-29-we-love-speed-2024/inp_charge.jpeg)
￼
Parmi les bonnes pratiques qu’on peut appliquer dès maintenant, et qui je dois le dire m’a paru un peu contre-intuitif :
Faire passer Babel sur les node_modules.
En fait, du point de vue d’un développeur, on peut se dire que cela va augmenter les temps de build drastiquement, et on
aurait sûrement raison. Mais en fait, il faut voir le bénéfice qu’il y a derrière. Si on personnalise les règles Babel

> pour qu’elles correspondent aux navigateurs de nos utilisateurs, on s’évite des transformations inutiles qui

Reviewer: ```suggestion
pour qu’elles correspondent aux navigateurs de nos utilisateurs, on évite des transformations inutiles qui
```

augmenteraient le poids de nos fichiers Javascript.

Une nouvelle fonctionnalité de React appelée RSC (React Server Components) permet combiner le rendu côté serveur avec
l'interactivité côté client.
Les RSC aident à réduire la taille du JavaScript dans le navigateur ce qui permet de réduire le temps d'interaction et

> de ce fait améliore grandement l'expérience utilisateur.

Reviewer: Pour éviter la répétition sur réduire : 

Les RSC aident à réduire la taille du Javascript dans le navigateur ce qui permet d’améliorer le temps d’interaction et donc l’expérience utilisateur. (A t’on besoin de le préciser ? 😁 )

Author: arf j'espère que nan

Vous l’avez deviné, c’est la partie "in-JS" qui pose un problème. Cela signifie que pour appliquer du style sur notre
site, c’est le Javascript qui s’en charge. Or dans un composant, comme le `<Footer />`, il y a beaucoup d’éléments et chacun va
avoir besoin son propre style. Si l’idée de colocaliser le CSS dans le JS n’est pas nocive en soi, le plus gros problème
était l'utilisation de [Styled-Components](https://styled-components.com/) qui calcule le style au moment du rendu, le rendant donc plus long.
FYI: Entre temps, nous avons chez Bedrock entamé une migration pour quitter Styled-Components au profit de [Linaria](https://linaria.dev/) pour le projet web et [Vanilla Extract](https://vanilla-extract.style/) pour le projet smart TV.

![Flamegraph du Footer de Bedrock](/images/posts/2024-10-29-we-love-speed-2024/flamgraph.jpeg)
￼
Autre information qui nous concerne chez Bedrock, au moment d’écrire ces lignes, nous sommes en train de mettre en
production la migration de React 17 vers React 18 sur le projet web.
D’après les retours d’expérience de Jean-Pierre, cette version de React aura un impact positif sur l’INP car il permet de faire moins de `render`.

> Enfin, Jean-Pierre nous laisse avec un ultime conseil pour que nos applications web soient pérennes : “Monitores (au

Reviewer: ```suggestion
Enfin, Jean-Pierre nous laisse avec un ultime conseil pour que nos applications web soient pérennes : “Monitore (au
```

> J'ai bien aimé ce talk ! J'ai trouvé que sa présentation était très accessible, il a su vulgariser des concepts et rendre un sujet chiant (la performance) intéressant 👏
> </div>

## Débogage des performances à l’aide des DevTools : Mise en pratique approfondie - [Umar Hansa](https://umaar.com/)

Avoir un œil sur ses performances, c’est essentiel pour pouvoir avancer dans la bonne direction et s’assurer qu’on
fournit à nos utilisateurs une expérience optimale. Fort heureusement pour nous les devs, on est bien lotis avec de très
bons outils. Il suffit d’ouvrir les Chrome DevTools pour s’en rendre compte. Il nous a été présenté lors de ce talk
comment bien utiliser les DevTools pour mesurer les performances de nos applications web et se mettre à la place de
nos utilisateurs. Par exemple, on peut brider son réseau et son CPU pour simuler une connexion 3G et un CPU lent. Dans
cette présentation, on nous a quand même rappelé que les DevTools ne sont pas une solution miracle, il est important de
tester sur de vrais devices pour le ressenti.

> Pour ce qui est de l'interprétation des données, une myriade d'outils sont à notre disposition pour nous aider à

Reviewer: ```suggestion
Pour ce qui est de l'interprétation des données, une myriade d'outils est à notre disposition pour nous aider à
```

comprendre ce que nous voyons. Par exemple, on peut ajouter des annotations dans le flamegraph comme des labels, des

> diagrammes ou encore des plages de temps. On peut aussi utiliser des custom tracks pour suivre des événements

Reviewer: `mettre en place des custom tracks` (pour éviter la répétition avec "utiliser" juste en dessous)

spécifiques. Au sein de notre application, on peut utiliser l’API User Timing pour ajouter des points de repère dans
notre code et ainsi mieux comprendre ce qui se passe au déclenchement d'événements spécifiques.

## Web Performance Testing - [Estela Franco](https://x.com/guaca)

On a aussi eu un talk sur l'intégration de Lighthouse, un outil de Google pour mesurer la performance de nos

> applications web, dans une CI. Cela permet de détecter les problèmes de performance avant qu'ils ne soient déployés en

Reviewer: `Nous avons également assisté à un talk sur l'intégration de Lighthouse, un outil de Google qui permet de mesurer la performance de nos applications web dans une CI.`

production. Il est possible de mettre des warnings, voire des erreurs empêchant de merger, si la performance de notre
application web ne respecte pas les standards que nous nous sommes fixés. L'idée est de s'assurer que la performance de

> notre application web est toujours au top et ne se dégrade pas dans le temps.

Reviewer: `si l’application ne respecte pas les standards que nous nous sommes fixés` (pour éviter la répétition de « performance de notre application web »)

performance avant qu'ils ne soient déployés en production. Il est possible de générer des warnings, voire des erreurs,
en cas de non-respect des standards de performance que nous avons définis. L'idée derrière la
mise en place de Lighthouse CI est de s'assurer que la performance de notre application web est toujours au
top et ne se dégrade pas dans le temps. Il est en effet plus facile de corriger une régression lorsqu'on peut savoir exactement quelle Pull Request ou commit l'a introduite.
![Key takeaways from the talk](/images/posts/2024-10-29-we-love-speed-2024/Key%20Takeaways.jpeg)

> <div style="display: flex">
> <img src="https://ca.slack-edge.com/T108ZKPMF-U01FQRQ8FT7-dfb12b21fb0d-192" alt="Julie" style="padding: 0;border-radius: 50%; height: 70px; margin: 10px">

> > Y'a moyen qu'on l'ajoute au projet smart TV, mais plus pour générer un rapport de performance quotidien plutôt que de le faire pour chaque push ou merge. 

Author: bien reçu 👌

chargent les différentes ressources nécessaires à l'affichage d'une page web. Plus spécifiquement, on nous a expliqué
comment les navigateurs interprètent le HTML pour déterminer quelles ressources charger en priorité. 

Le talk était très intéressant, mais la conclusion est un peu frustrante : il est pour le moment impossible de prévoir
l'ordre de chargement des ressources par le navigateur à partir du même HTML. En effet, les navigateurs ont des
comportements différents entre eux et même par version 🤯. Chrome a, par exemple, un comportement très différent cette année
par rapport à deux ans en arrière.

Même si on est tenté de vouloir contrôler le chargement des ressources, il est important de se rappeler que le navigateur
est très bien optimisé pour charger les ressources de manière efficace. Il est donc préférable de laisser le navigateur
faire son travail plutôt que de vouloir le contrôler. L'attribut `preload` est un bon exemple de ce que l'on peut faire
pour aider le navigateur à charger les ressources de manière plus efficace. Il faut cependant l'utiliser avec parcimonie
et de manière chirurgicale pour ne pas interférer avec le travail du navigateur.
￼

> ![Preload with surgical precision](/images/posts/2024-10-29-we-love-speed-2024/preload_surgical.jpeg)

Reviewer 3: J'ai le même problème (aussi sur Firefox)

> ![Preload with surgical precision](/images/posts/2024-10-29-we-love-speed-2024/preload_surgical.jpeg)

Author: je crois que c'est lié à un import des images depuis l'appli notes d'apple, je les vois pas dans mon IDE c'est un peu relou

applications web, mais elles peuvent aussi être une source de problèmes de performance. En effet, les polices de
caractères peuvent être très lourdes et ralentir le chargement de nos pages. Il est donc important de bien les choisir
et de les optimiser pour garantir une bonne performance. Il existe plusieurs techniques pour optimiser les polices,
notamment en réalisant un subset de la police pour ne télécharger que les glyphes dont on a besoin. (En français, on a
besoin que de 165 glyphes comparé à 528 pour le latin).
Il existe des outils pour nous aider à réaliser ces subsets comme: [Font Subsetter](https://everythingfonts.com/subsetter), [fontTools](https://fonttools.readthedocs.io/) ou
[Glyphanger](https://www.zachleat.com/web/glyphhanger/) .
> <div style="display: flex">
> Attention à ne pas abuser des subsets, car cela peut entraîner des problèmes de lisibilité du texte. Le fameux t🠉fu .
> <img src="/images/avatar/j_poissonnet.jpg" alt="Jules" style="padding: 0;border-radius: 50%; height: 70px; margin: 10px">
> </div>

![Say no to tofu](/images/posts/2024-10-29-we-love-speed-2024/tofu.jpeg)

> Il est aussi possible de minimiser le nombre de fichiers en utilisant des polices variables. Un bon exemple, c'est la police

Reviewer: ```suggestion
Il est aussi possible de minimiser le nombre de fichiers en utilisant des polices variables. Un bon exemple est la police
```

---
layout: post
title: We love speed 2024 ❤️
description: Retour sur la conférence We love speed 2024 ❤️, une conférence annuelle axée sur la performance du web.
author: [ j_nginn, j_poissonnet ]
tags: [ performance, conference, webperf, javascript, react, web, frontend ]
color: rgb(251,87,66)
language: fr
thumbnail: "images/posts/2024-10-29-we-love-speed-2024/welovespeed_2024-1709240237.jpg"
---

Nous avons eu la chance de participer à la conférence We love speed, une conférence annuelle axée sur la
performance du web. C’est un domaine qui nous passionne et nous sommes très content d’avoir pu y assister.
Le thème de cette édition, c’est l’INP. En effet, cette métrique de performance a été ajoutée aux core web vitals par

> [Google _récemment_](https://developers.google.com/search/blog/2023/05/introducing-inp?hl=fr)

Reviewer 3: ```suggestion
[Google _récemment_](https://developers.google.com/search/blog/2023/05/introducing-inp?hl=fr).
```

L’objectif de cette métrique est de refléter l’expérience utilisateur en mesurant la réactivité d’une application.
Elle observe le temps entre une action utilisateur et une réponse visuelle de notre interface.

![L'équipe frontend à la we love speed](/images/posts/2024-10-29-we-love-speed-2024/team_picture.jpeg)

## HTMX, le nouvel atout pour vos projets SSR - [Ewen Quimerc‘h](https://ewen.quimerch.com/)

Lors de ce talk, nous avons découvert un outil très intéressant. Il s'agit de HTMX: une bibliothèque Javascript qui
permet d'ajouter des fonctionnalités de type SPA (Single Page Application) à une application web classique et de façon
non intrusive. Par exemple, on peut surcharger les liens `<a>` pour qu’ils chargent une nouvelle page en AJAX grâce à un
attribut ajouté au HTML. Ce mode de fonctionnement est très intéressant, car il permet de garder une application web

> classique avec tous ces comportements le temps que HTMX se charge. C’est-a-dire que si HTMX venait à ne pas démarrer,

Reviewer 3: Si vous parlez de l'application et de ses comportements
```suggestion
classique avec tous ses comportements le temps que HTMX se charge. C’est-a-dire que si HTMX venait à ne pas démarrer,
```

votre application web se comporterait de la même manière, mais sans les améliorations de temps d’interaction.

HTMX surcharge la manière dont vos liens et images vont être chargés par le navigateur.
Ainsi, lors de la prochaine interaction, ce dernier sera déjà prêt à servir les ressources.
Le principe de HTMX consiste à ajouter des balises HTML spécifiques dans le DOM qu'il va lire et en déduire les
comportements à son chargement.

> Cette manipulation est appelé le "DOM morphing". Grâce à ce processus le temps de chargement est réduit et on évite

Reviewer 4: appelée
